drivers/i2s/i2s_ll_stm32.c

```diff
@@ -0,0 +1,1171 @@
+	}
+
+	__ASSERT_NO_MSG(stream->mem_block != NULL);
+
+	/* All block data sent */
+	k_mem_slab_free(stream->cfg.mem_slab, &stream->mem_block);
+	stream->mem_block = NULL;
+
+	/* Stop transmission if there was an error */
+	if (stream->state == I2S_STATE_ERROR) {
+		SYS_LOG_ERR("TX error detected");
+		goto tx_disable;
+	}
+
+	/* Stop transmission if we were requested */
+	if (stream->last_block) {
+		stream->state = I2S_STATE_READY;
+		goto tx_disable;
+	}
+
+	/* Prepare to send the next data block */
+	ret = queue_get(&stream->mem_block_queue, &stream->mem_block,
+			&mem_block_size);
+	if (ret < 0) {
+		if (stream->state == I2S_STATE_STOPPING) {
+			stream->state = I2S_STATE_READY;
+		} else {
+			stream->state = I2S_STATE_ERROR;
+		}
+		goto tx_disable;
+	}
+	k_sem_give(&stream->sem);
+
+	/* Assure cache coherency before DMA read operation */
+	DCACHE_CLEAN(stream->mem_block, mem_block_size);
+
+	ret = start_dma(dev_data->dev_dma, stream->dma_channel,
+			&stream->dma_cfg,
+			stream->mem_block,
+			(void *)LL_SPI_DMA_GetRegAddr(cfg->i2s),
+			stream->cfg.block_size);
+	if (ret < 0) {
+		SYS_LOG_DBG("Failed to start TX DMA transfer: %d", ret);
+		goto tx_disable;
+	}
+
+	return;
+
+tx_disable:
+	tx_stream_disable(stream, dev);
+}
+
+static u32_t i2s_stm32_irq_count;
+static u32_t i2s_stm32_irq_ovr_count;
+
+static void i2s_stm32_isr(void *arg)
+{
+	struct device *const dev = (struct device *) arg;
+	const struct i2s_stm32_cfg *cfg = dev->config->config_info;
+	struct i2s_stm32_data *const dev_data = DEV_DATA(dev);
+	struct stream *stream = &dev_data->rx;
+
+	SYS_LOG_ERR("%s: err=%d", __func__, LL_I2S_ReadReg(cfg->i2s, SR));
+	stream->state = I2S_STATE_ERROR;
+
+	/* OVR error must be explicitly cleared */
+	if (LL_I2S_IsActiveFlag_OVR(cfg->i2s)) {
+		i2s_stm32_irq_ovr_count++;
+		LL_I2S_ClearFlag_OVR(cfg->i2s);
+	}
+
+	i2s_stm32_irq_count++;
+}
+
+static int i2s_stm32_initialize(struct device *dev)
+{
+	const struct i2s_stm32_cfg *cfg = DEV_CFG(dev);
+	struct i2s_stm32_data *const dev_data = DEV_DATA(dev);
+	int ret, i;
+
+	/* Enable I2S clock propagation */
+	ret = i2s_stm32_enable_clock(dev);
+	if (ret < 0) {
+		SYS_LOG_ERR("%s: clock enabling failed: %d",  __func__, ret);
+		return -EIO;
+	}
+
+	cfg->irq_config(dev);
+
+	k_sem_init(&dev_data->rx.sem, 0, CONFIG_I2S_STM32_RX_BLOCK_COUNT);
+	k_sem_init(&dev_data->tx.sem, CONFIG_I2S_STM32_TX_BLOCK_COUNT,
+		   CONFIG_I2S_STM32_TX_BLOCK_COUNT);
+
+	for (i = 0; i < STM32_DMA_NUM_CHANNELS; i++) {
+		active_dma_rx_channel[i] = NULL;
+		active_dma_tx_channel[i] = NULL;
+	}
+
+	/* Get the binding to the DMA device */
+	dev_data->dev_dma = device_get_binding(dev_data->dma_name);
+	if (!dev_data->dev_dma) {
+		SYS_LOG_ERR("%s device not found", dev_data->dma_name);
+		return -ENODEV;
+	}
+
+	SYS_LOG_INF("%s inited", dev->config->name);
+
+	return 0;
+}
+
+static int rx_stream_start(struct stream *stream, struct device *dev)
+{
+	const struct i2s_stm32_cfg *cfg = dev->config->config_info;
```

Comment:
Since we use DEV_DATA macro, shouldn't we use here the DEV_CFG one? Like in `i2s_stm32_initialize` function. Several other functions are also affected.

Reply:
Oooops.... correct!
Thanks for catching it out.

Reply:
Thanks for fixing, however there are still a couple of functions which don't use DEV_CFG macro.

Reply:
sigh... I'm doing too many stuff at the same time.
Thanks.